plugins/com.google.cloud.tools.eclipse.jdt.launching/src/com/google/cloud/tools/eclipse/jdi/internal/connect/SocketConnection.java

```diff
@@ -0,0 +1,154 @@
+/*******************************************************************************
+ * Copyright (c) 2000, 2016 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *     Google Inc - add support for accepting multiple connections
+ *******************************************************************************/
+package com.google.cloud.tools.eclipse.jdi.internal.connect;
+
+import com.sun.jdi.connect.spi.ClosedConnectionException;
+import com.sun.jdi.connect.spi.Connection;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.Socket;
+
+/**
+ * A fork of {@link org.eclipse.jdi.internal.connect.SocketConnection} that
+ * stores the socket, input stream, and output stream.
+ */
+public class SocketConnection extends Connection {
+
+	// for attaching connector
+	private Socket fSocket;
+
+	private InputStream fInput;
+
+	private OutputStream fOutput;
+
+	SocketConnection(Socket socket, InputStream in, OutputStream out) {
+		fSocket = socket;
+		fInput = in;
+		fOutput = out;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * 
+	 * @see com.sun.jdi.connect.spi.Connection#close()
+	 */
+	@Override
+	public synchronized void close() throws IOException {
+		if (fSocket == null)
+			return;
+
+		fSocket.close();
+		fSocket = null;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * 
+	 * @see com.sun.jdi.connect.spi.Connection#isOpen()
+	 */
+	@Override
+	public synchronized boolean isOpen() {
+		return fSocket != null;
```

Comment:
is it possible that due to some exception (remote host closes connection), the socket is actually closed/in error state, but not null? (Normally it's null as the result of the `close()` method)